Add two levels to use_c10_dispatcher

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26434 Move Generator ops to c10
* #26273 Enable creation of boxing wrappers for some aten operators
* **#26272 Add two levels to use_c10_dispatcher**
* #26337 c10::KernelFunction

```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

smessmer committed:
Add two levels to use_c10_dispatcher

```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

dzhulgakov commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -1598,28 +1595,28 @@
 
 - func: quantized_max_pool2d(Tensor self, int[2] kernel_size, int[2] stride=[], int[2] padding=0, int[2] dilation=1) -> Tensor
-  use_c10_dispatcher: True
+  use_c10_dispatcher: 'unboxed_only'
```
nit: I don't think you need quotes in yaml

dzhulgakov reviewed (approved):
It looks ok to me

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

smessmer committed:
Update on "Add two levels to use_c10_dispatcher"


```
use_c10_dispatcher: 'unboxed_only'
```
This is the previous implementation. The operator is registered with c10, but only in its unboxed form. No boxing wrapper is generated.

```
use_c10_dispatcher: 'full'
```
This does everything done by 'unboxed_only', but additionally creates a boxing wrapper so the op can be called through the c10 dispatcher using a boxed operator call.

This only changes registration, not the calling path. These operators are still called through the unboxed function pointer.

The final goal is to have 'full' for all operators, but this isn't immediately going to work for all ops.

[namedtensor ci]

Differential Revision: [D17393317](https://our.internmc.facebook.com/intern/diff/D17393317/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -157,6 +169,17 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+#else
+    static c10::OperatorHandle op = c10::Dispatcher::singleton().findSchema({"aten::${name}", "${overload_name}"}).value();
+    return c10::Dispatcher::singleton().callUnboxed<${formals_types_with_return}>(
+        op, impl::dispatchTypeId(type_set())${method_actuals_with_comma_prefix});
```
This gonna need to do the multiple dispatch thing.

smessmer replied:
ouch I think this got screwed up by a merge conflict with the revert of the multi dispatch diff.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -190,6 +213,19 @@ def TypedDict(name, attrs, total=True):  # type: ignore
+    static c10::OperatorHandle op = c10::Dispatcher::singleton()
+        .findSchema({"aten::${name}", "${overload_name}"}).value();
+    return c10::Dispatcher::singleton().callUnboxed<${formals_types_with_return}>(
+        op, impl::dispatchTypeId(${inferred_type_set})${native_actuals_with_comma_prefix});
```
Also going to need multiple dispatch bits too.

smessmer replied:
this is using `inferred_type_Set`, same as the aten version, so should be ok, right?

Ultimately we want to move this logic into the dispatcher, but I'm waiting for more of your dispatch diffs to land before doing that because it would likely conflict.

ezyang replied:
Multidispatch is landed, so all of the stuff you need is landed already.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1144,7 +1180,12 @@ def gen_tensor_method(option):
+                method_definition = TENSOR_METHOD_DEFINITION
+            elif option['use_c10_dispatcher'] == 'unboxed_only':
+                method_definition = C10_UNBOXEDONLY_TENSOR_METHOD_DEFINITION
+            else:
```
Can we add an `assert['use_c10_dispatcher'] == 'full'` here? (or whatever that value is)

smessmer replied:
There is an assertion in `native_parse.yaml` but I'll also add one here.

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1185,9 +1226,12 @@ def gen_namespace_function(option, dispatch_tensor, dispatch_options):
                     option, static_dispatch_function_body=static_dispatch_function_body)
             else:
-                if not option['use_c10_dispatcher']:
+                if (option['use_c10_dispatcher'] == 'no'):
```
You don't need parentheses here

smessmer replied:
nope

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1185,9 +1226,12 @@ def gen_namespace_function(option, dispatch_tensor, dispatch_options):
+                elif option['use_c10_dispatcher'] == 'unboxed_only':
+                    fn_definition = C10_UNBOXEDONLY_FUNCTION_DEFINITION.substitute(
+                        option, static_dispatch_function_body=static_dispatch_function_body)
                 else:
```
assert here plz

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1279,8 +1323,12 @@ def add_namedtensor_enabled_macro(code):
+            if (option['use_c10_dispatcher'] == 'full'):
+                top_env['function_registrations'].append(
+                    check_namedtensor_enabled(DEFAULT_FUNCTION_REGISTRATION.substitute(option)))
+            else:
```
assert here

ezyang commented on aten/src/ATen/function_wrapper.py:
```diff
@@ -1743,8 +1791,12 @@ def process_native(option):
+                    if (option['use_c10_dispatcher'] == 'full'):
+                        function_registrations.append(
+                            BACKEND_FUNCTION_REGISTRATION.substitute(env))
+                    else:
```
assert here

ezyang reviewed (approved):
There is nothing obviously wrong with this patch, but this patch adds **four** more blocks of generated code to `function_wrapper.py`. As I have commented in the patch earlier in this stack, I think at least one of these is unnecessary: the basic API should work whether or not the boxing code works or not. So you shouldn't have to switch between these cases in the C10 dispatcher. When you want to enforce boxing works on everything, you just remove the sfinae guard in C10 side.

smessmer commented:
@ezyang saw your comments after landing it. Created a new PR for addressing them: https://github.com/pytorch/pytorch/pull/26587
